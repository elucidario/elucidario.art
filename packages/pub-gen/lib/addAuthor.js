import path from "path";
import { getPaths, createFile } from "@elucidario/pkg-paths";
import { Console } from "@elucidario/pkg-console";

import { pubGenPrompt } from "./prompt/pubGenPrompt.js";
import {
    makeInquirer,
    pubGenConfig,
    packageJson,
    cleanFalsy,
} from "./utils.js";

let authors = [];

export const addAuthor = async (args) => {
    const { publication } = args;
    const pkg = packageJson(publication);
    const console = new Console(pkg);

    await makeInquirer({
        title: "Adicionando autor...",
        publication,
        prompt: pubGenPrompt("addAuthor", args),
        callback: async (answers) => {
            const { contributor, addMoreAuthor } = cleanFalsy(answers);

            authors.push(contributor);

            if (addMoreAuthor) {
                console.success({
                    message: `Autor ${contributor.title} adicionado com sucesso! Adicione mais autores:`,
                });
                addAuthor(args);
            } else {
                const pubGenJson = pubGenConfig(publication);

                pubGenJson.contributors = [
                    ...pubGenJson.contributors,
                    ...authors,
                ];

                createFile(
                    {
                        filePath: path.resolve(
                            getPaths().publications,
                            publication,
                            "pub-gen.json"
                        ),
                        ext: "json",
                    },
                    pubGenJson
                );

                const authorsNames = authors
                    .map((author) => author.title)
                    .join(", ");

                console.success({
                    message: `Autores adicionados com sucesso: ${authorsNames}`,
                });

                authors = [];
            }
        },
    });
};
