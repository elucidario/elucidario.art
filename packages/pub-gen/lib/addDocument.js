"use strict";

import fs from "fs";
import path, { dirname } from "path";
import inquirer from "inquirer";
import chalk from "chalk";

import lodash from "lodash";

import { createDoc } from "@elucidario/pkg-md-to-gdoc";

import { pubGenPrompt } from "./prompt.js";
import { getCredentials } from "./getCredentials.js";

const { merge } = lodash;

const __dirname = path.resolve();

export const addDocument = async (args) => {
    const rootPath = path.resolve(__dirname, "..", "..");
    const templatePath = path.resolve(
        rootPath,
        "packages",
        "pub-gen",
        "template"
    );
    const publication = args.publication;
    const pubPath = path.resolve(rootPath, "publications");
    const dirName = path.resolve(pubPath, publication);

    inquirer.prompt(pubGenPrompt("addDocument", args)).then(async (answers) => {
        const { anotherDocument, document } = answers;

        let documents = [];
        if (anotherDocument) {
            console.log(
                chalk.green(`${document.title} adicionado com sucesso`)
            );
            documents = await createGoogleIntegration(
                document,
                templatePath,
                dirName,
                rootPath
            );
            addDocument(args);
        } else {
            documents = await createGoogleIntegration(
                document,
                templatePath,
                dirName,
                rootPath
            );
            /**
             * Save pub-gen.json
             */
            saveDocumentsToPubGenJson(documents, dirName);
        }
    });
};

const saveDocumentsToPubGenJson = (documents, dirName) => {
    const pubGenJson = JSON.parse(
        fs.readFileSync(path.resolve(dirName, "pub-gen.json"))
    );
    pubGenJson.documents = [...pubGenJson.documents, ...documents];
    try {
        fs.writeFileSync(
            path.resolve(dirName, "pub-gen.json"),
            JSON.stringify(pubGenJson, null, 4)
        );
        const documentsNames = documents.map((doc) => doc.title).join(", ");
        console.log(
            chalk.green(`Successfully create documents: ${documentsNames}`)
        );
    } catch (error) {
        console.log(chalk.red("It was not possible to save pub-gen.json"));
        console.error(error);
    }
};

const documents = [];
const createGoogleIntegration = async (document, from, to, rootPath) => {
    try {
        /**
         * Try to get credentials
         */
        let credentials = {};
        try {
            credentials = getCredentials();
        } catch (error) {
            console.log(chalk.red("It was not possible to get credentials"));
            console.error(error);
            return error;
        }

        switch (document.type) {
            case "doc":
                /**
                 * Create references directory and copy demo references
                 */
                createReferences(document.preset, from, to);

                /**
                 * Create content directory and copy demo content
                 */
                createDemoContent(document.language, from, to);

                /**
                 * Create Google Docs
                 */
                try {
                    const defaultDoc = {
                        ...document,
                        keywords: document.keywords
                            .split(",")
                            .map((keyword) => keyword.trim()),
                    };
                    const response = await createDoc(
                        document.title,
                        credentials,
                        rootPath
                    );
                    const doc = merge({}, defaultDoc, response);
                    documents.push(doc);
                    return documents;
                } catch (error) {
                    console.log(
                        chalk.red("It was not possible to create Google Docs")
                    );
                    console.error(error);
                    return error;
                }

            case "slide":
                console.warn(chalk.yellowBright("Slides not implemented yet"));
                // docs = await createDoc(
                //     document.title,
                //     credentials,
                //     rootPath
                // );
                // console.log(docs);
                throw new Error("Slides not implemented yet");

            case "sheet":
                console.warn(chalk.yellowBright("Sheets not implemented yet"));
                // docs = await createDoc(
                //     document.title,
                //     credentials,
                //     rootPath
                // );
                // console.log(docs);
                throw new Error("Sheets not implemented yet");

            default:
                throw new Error("Document type not found");
        }
    } catch (error) {
        console.log(
            chalk.red("It was not possible to create integration with Google")
        );
        console.error(error);
        return error;
    }
};

const createReferences = (preset, from, to) => {
    try {
        let demoRefs = fs.readdirSync(path.resolve(from, "references", preset));
        if (demoRefs.length > 0) {
            fs.mkdirSync(path.resolve(to, "references", preset), {
                recursive: true,
            });
            demoRefs.forEach((file) => {
                fs.copyFileSync(
                    path.resolve(from, "references", preset, file),
                    path.resolve(to, "references", preset, file)
                );
            });
        }
        console.log(chalk.green("Successfully created references."));
    } catch (error) {
        console.log(chalk.red("It was not possible to create references"));
        console.error(error);
    }
};

const createDemoContent = (language, from, to) => {
    try {
        const demoContent = fs.readdirSync(path.resolve(from, "content"));
        fs.mkdirSync(path.resolve(to, "content", language), {
            recursive: true,
        });
        demoContent.forEach((file) => {
            fs.copyFileSync(
                path.resolve(from, "content", file),
                path.resolve(to, "content", language, file)
            );
        });
        console.log(chalk.green("Successfully created demo content."));
    } catch (error) {
        console.log(chalk.red("It was not possible to create demo content"));
        console.error(error);
    }
};
