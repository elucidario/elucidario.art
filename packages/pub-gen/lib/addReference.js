import {
    authorPrompt,
    datePrompt,
    referencePrompt,
    authorProperties,
    dateProperties,
} from "./prompt/referencePrompt.js";
import inquirer from "inquirer";

import { Console } from "@elucidario/pkg-console";
import packageJson from "../package.json" assert { type: "json" };
import { kebabCase } from "lodash-es";

const console = new Console(packageJson);

export const addReference = async (defaults) => {
    try {
        console.log("Selecione o tipo de referencia:");
        inquirer
            .prompt(referencePrompt("create", defaults))
            .then(async (answers) => {
                await addReferenceByType(answers.type, defaults);
            });
    } catch (error) {
        console.error(error);
    }
};

export const addReferenceByType = async (type, defaults) => {
    try {
        console.log(`Adicionando nova referencia do tipo ${type}`);
        await inquirer
            .prompt(referencePrompt(type, defaults))
            .then(async (answers) => {
                // parse answers and return a new object with only the non-empty values
                let reference = Object.keys(answers[type]).reduce(
                    (acc, cur) => {
                        if (answers[type][cur].length) {
                            acc[cur] = answers[type][cur];
                        }
                        return acc;
                    },
                    {}
                );
                reference["type"] = type;
                reference["_slug"] = kebabCase(reference.title);

                // configura a referencia
                try {
                    // configura os autores
                    try {
                        console.log("Selecione o tipo de autor:");
                        await inquirer
                            .prompt(authorPrompt("type", { defaults }))
                            .then(async (answers) => {
                                const { type } = answers;

                                // configura os nomes dos autores
                                try {
                                    const authors = await author(
                                        type,
                                        defaults
                                    ).then((authorNames) => {
                                        // parse authorNames and return a new object with only the non-empty values
                                        return Object.keys(authorNames).reduce(
                                            (acc, cur) => {
                                                if (authorNames[cur].length) {
                                                    acc[cur] = authorNames[cur];
                                                }
                                                return acc;
                                            },
                                            {}
                                        );
                                    });

                                    reference = { ...reference, ...authors };
                                } catch (error) {
                                    console.error(error);
                                    throw new Error(error);
                                }
                            });
                    } catch (error) {
                        console.error(error);
                        throw new Error(error);
                    }

                    // configura as datas
                    try {
                        console.log("Selecione o tipo de data:");
                        await inquirer
                            .prompt(datePrompt("type", { defaults }))
                            .then(async (answers) => {
                                const { type } = answers;

                                // configura os tipos de datas
                                try {
                                    const dates = await date(
                                        type,
                                        defaults
                                    ).then((dateNames) => {
                                        // parse authorNames and return a new object with only the non-empty values
                                        return Object.keys(dateNames).reduce(
                                            (acc, cur) => {
                                                if (
                                                    Object.keys(dateNames[cur])
                                                        .length
                                                ) {
                                                    acc[cur] = dateNames[cur];
                                                }
                                                return acc;
                                            },
                                            {}
                                        );
                                    });

                                    reference = { ...reference, ...dates };
                                } catch (error) {
                                    console.error(error);
                                    throw new Error(error);
                                }
                            });
                    } catch (error) {
                        console.error(error);
                        throw new Error(error);
                    }

                    // configura as propriedades comuns
                    try {
                        console.log("Selecione as propriedades comuns:");
                        await inquirer
                            .prompt(
                                referencePrompt("common", {
                                    defaults,
                                })
                            )
                            .then(async (answers) => {
                                const cleanAnswers = Object.keys(
                                    answers
                                ).reduce((acc, cur) => {
                                    if (answers[cur].length) {
                                        acc[cur] = answers[cur];
                                    }
                                    return acc;
                                }, {});

                                reference = {
                                    ...reference,
                                    ...cleanAnswers,
                                };
                            });
                    } catch (error) {
                        console.error(error);
                        throw new Error(error);
                    }

                    // configura as propriedades de arquivo
                    try {
                        console.log("Selecione as propriedades de arquivo:");
                        await inquirer
                            .prompt(
                                referencePrompt("archive", {
                                    defaults,
                                })
                            )
                            .then(async (answers) => {
                                const cleanAnswers = Object.keys(
                                    answers
                                ).reduce((acc, cur) => {
                                    if (answers[cur].length) {
                                        acc[cur] = answers[cur];
                                    }
                                    return acc;
                                }, {});

                                reference = {
                                    ...reference,
                                    ...cleanAnswers,
                                };
                            });
                    } catch (error) {
                        console.error(error);
                        throw new Error(error);
                    }
                } catch (error) {
                    console.error(error);
                    throw new Error(error);
                }

                console.log({ reference }, true);
            });
    } catch (error) {
        throw new Error(error);
    }
};

// cria um objeto em que cada propriedade é um item do array authorProperties
const authorNames = authorProperties.reduce((acc, cur) => {
    acc[cur] = [];
    return acc;
}, {});

export const author = async (type, { defaults }) => {
    console.log(`Adicionando novo autor do tipo ${type}`);
    await inquirer
        .prompt(
            authorPrompt("names", {
                defaults,
                name: type,
            })
        )
        .then(async (answers) => {
            const { addMore, ...definedAuthor } = answers;
            const cleanAuthor = Object.keys(definedAuthor[type]).reduce(
                (acc, cur) => {
                    if (definedAuthor[type][cur]) {
                        acc[cur] = definedAuthor[type][cur];
                    }
                    return acc;
                },
                {}
            );
            authorNames[type].push(cleanAuthor);
            if (addMore) {
                await inquirer
                    .prompt(
                        authorPrompt("type", {
                            defaults,
                        })
                    )
                    .then(async (answers) => {
                        const { type } = answers;
                        await author(type, { defaults });
                    });
            }
        });

    return authorNames;
};

// cria um objeto em que cada propriedade é um item do array dateProperties
const dateNames = dateProperties.reduce((acc, cur) => {
    acc[cur] = {};
    return acc;
}, {});
export const date = async (type, { defaults }) => {
    console.log(`Adicionando nova data do tipo ${type}`);
    await inquirer
        .prompt(
            datePrompt("date", {
                defaults,
                name: type,
            })
        )
        .then(async (answers) => {
            const { addMore, ...definedDate } = answers;

            const cleanDate = Object.keys(definedDate[type]).reduce(
                (acc, cur) => {
                    if (definedDate[type][cur]) {
                        acc[cur] = definedDate[type][cur];
                    }
                    return acc;
                },
                {}
            );
            dateNames[type] = cleanDate;
            if (addMore) {
                await inquirer
                    .prompt(
                        datePrompt("type", {
                            defaults,
                            name: type,
                        })
                    )
                    .then(async (answers) => {
                        const { type } = answers;
                        await date(type, { defaults, name: type });
                    });
            }
        });

    return dateNames;
};
