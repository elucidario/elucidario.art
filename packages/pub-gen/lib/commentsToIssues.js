import { comments } from "@elucidario/pkg-g-api-utils";
import path from "path";
import fs from "fs";
import {
    toMD,
    heading,
    table,
    italic,
    bold,
} from "@elucidario/pkg-docusaurus-md";

import { Console } from "@elucidario/pkg-console";
import { getPaths } from "@elucidario/pkg-paths";

import { getCredentials } from "./getCredentials.js";
import { githubAuthenticate } from "./githubAuthenticate.js";

const paths = getPaths();

export const createIssue = async ({
    issue,
    octokit,
    fileId,
    commentId,
    drive,
}) => {
    const { data } = await octokit.rest.issues.create({
        owner: "hgodinho",
        repo: "elucidario",
        ...issue,
    });

    const reply = await drive.replyComment(
        fileId,
        commentId,
        `[pub-gen] issue: ${data.html_url}`
    );

    console.log(
        {
            reply: {
                created: reply.data.createdDate,
                data: reply.data.content,
            },
            issueUrl: data.html_url,
            issueTitle: data.title,
        },
        {
            title: `Issue criada: ${data.html_url}`,
            type: "success",
            defaultLog: true,
        }
    );
};

export const mountIssue = ({ fileId, comment, pkgName, publication }) => {
    return {
        title: `Comentário em \`${comment.fileTitle}\` - ${comment.htmlContent}`,
        body: toMD([
            heading(
                1,
                `Comentário no arquivo \`${comment.fileTitle}\` de \`${pkgName}\``
            ),
            heading(
                6,
                `por: ${comment.author.displayName} em: ${new Date(
                    comment.modifiedDate
                ).toLocaleString()}`
            ),
            italic(
                `[link para o comentário](https://docs.google.com/document/d/${fileId}/edit?disco=${comment.commentId})`
            ),
            table({
                headers: ["fileId", "fileTitle", "commentId", "status"],
                rows: [
                    [
                        fileId,
                        comment.fileTitle,
                        comment.commentId,
                        comment.status,
                    ],
                ],
            }),
            comment.context
                ? toMD([
                      heading(2, "Contexto do comentário"),
                      `${bold("type:")} ${comment.context.type}`,
                      `> ${comment.context.value}`,
                  ])
                : "",
            heading(2, "Conteúdo do comentário"),
            comment.htmlContent,
            "---",
            italic(
                `issue criada automaticamente por [pub-gen](https://elucidario.art/pub-gen)`
            ),
        ]),
        labels: ["project", `pub-${publication}`],
    };
};

export const commentsToIssues = async ({ fileId, publication }) => {
    if (!publication) {
        throw new Error("Publication not defined");
    }
    if (!fileId) {
        throw new Error("FileId not defined");
    }

    const pkg = JSON.parse(
        fs.readFileSync(
            path.resolve(paths.publications, publication, "package.json")
        )
    );

    const console = new Console(pkg);

    const credentials = getCredentials();

    const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

    const octokit = await githubAuthenticate();

    const commentList = await comments({
        fileId,
        credentials,
        tokenPath: paths.root,
        scopes: ["https://www.googleapis.com/auth/drive"],
        options: {
            maxResults: 100,
        },
        callback: async (comments, drive, nextPageToken) => {
            // filter comments
            comments = comments.filter((comment) => {
                // if resolved, ignore
                if (comment.status === "resolved") {
                    return false;
                }

                // if reply includes pub-gen, issue already created, ignore
                let reply = false;
                comment.replies.map((rep) => {
                    if (
                        "content" in rep &&
                        rep.content.includes("[pub-gen] issue: ")
                    ) {
                        reply = true;
                    }
                });
                if (reply) {
                    return false;
                }

                // if open, create issue
                return comment.status === "open";
            });

            console.log(`total: ${comments.length}`);

            comments.map(async (comment) => {
                const issue = mountIssue({
                    fileId,
                    comment,
                    pkgName: pkg.name,
                    publication,
                });

                await delay(1000).then(async () => {
                    await createIssue({
                        issue,
                        octokit,
                        fileId,
                        commentId: comment.commentId,
                        drive,
                    });
                });
            });
        },
    });
};
