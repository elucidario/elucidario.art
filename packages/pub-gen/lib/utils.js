import path from "path";
import { getPaths, readFile } from "@elucidario/pkg-paths";
import { toMD } from "@elucidario/pkg-docusaurus-md";
import inquirer from "inquirer";
import { Console } from "@elucidario/pkg-console";
import { kebabCase } from "lodash-es";
import { fromMarkdown } from "mdast-util-from-markdown";

const pkg = readFile({
    filePath: path.resolve(getPaths().packages, "pub-gen", "package.json"),
}).value;

/**
 * Get publication config
 * @param {string} publication
 * @returns {Object}
 * @throws {Error}
 */
export function pubGenConfig(publication) {
    try {
        return readFile({
            filePath: path.resolve(
                getPaths().publications,
                publication,
                "pub-gen.json",
            ),
            ext: "json",
        }).value;
    } catch (error) {
        throw new Error(
            `Não foi possível encontrar o pub-gen.json da publicação ${publication}: ${error}`,
        );
    }
}

/**
 * Get publication package.json
 * @param {string} publication
 * @returns {Object}
 * @throws {Error}
 */
export function packageJson(publication) {
    try {
        return readFile({
            filePath: path.resolve(
                getPaths().publications,
                publication,
                "package.json",
            ),
            ext: "json",
        }).value;
    } catch (error) {
        throw new Error(
            `Não foi possível encontrar o package.json da publicação ${publication}: ${error}`,
        );
    }
}

/**
 * Get reference index of a publication
 *
 * @param {string} publication
 * @returns {Object}
 * @throws {Error}
 */
export function referenceIndex(publication) {
    try {
        return readFile({
            filePath: path.resolve(
                getPaths().publications,
                publication,
                "references",
                "index.json",
            ),
            ext: "json",
        }).value;
    } catch (error) {
        throw new Error(
            `Não foi possível encontrar o index das referências da publicação ${publication}: ${error}`,
        );
    }
}

/**
 * Get references of a publication
 *
 * @param {string} publication
 * @returns {Array}
 * @throws {Error}
 */
export function referencesFrom(publication) {
    try {
        return referenceIndex(publication).items.map((item) => {
            return readFile({
                filePath: path.resolve(
                    getPaths().references,
                    item.path.replace("<references>/", ""),
                ),
                ext: "json",
            }).value;
        });
    } catch (error) {
        throw new Error(
            `Não foi possível encontrar as referências da publicação ${publication}: ${error}`,
        );
    }
}

/**
 * Generate a debounced function
 *
 * @param {function} func
 * @param {number} delay
 * @returns {function}
 */
export async function debounce(func, delay) {
    let timeoutId;
    return async (...args) => {
        clearTimeout(timeoutId);
        timeoutId = setTimeout(async () => {
            await func(...args);
        }, delay);
    };
}

/**
 * Clean falsy values from objects
 *
 * @param {object} obj
 * @returns {object}
 */
export function cleanFalsy(obj) {
    return Object.fromEntries(
        Object.entries(obj).reduce((acc, [key, value]) => {
            if (value !== null && value !== undefined && value !== "") {
                if (
                    typeof value === "object" &&
                    !Array.isArray(value) &&
                    Object.keys(value).length > 0
                ) {
                    value = cleanFalsy(value);
                    acc.push([key, value]);
                } else if (Array.isArray(value) && value.length) {
                    value = value.map((item) => {
                        if (typeof item === "object") {
                            return cleanFalsy(item);
                        }
                        return item;
                    });
                    acc.push([key, value]);
                } else if (typeof value === "string" && value.length) {
                    value = value.trim();
                    acc.push([key, value]);
                } else if (typeof value === "boolean" || value === 0) {
                    acc.push([key, value]);
                } else {
                    return acc;
                }
            }
            return acc;
        }, []),
    );
}

/**
 * Get publication package name
 *
 * @param {string} publication
 * @returns {string}
 */
export const publicationPkgName = (publication) =>
    `@elucidario/pub-${kebabCase(publication.trim())}`;

/**
 * Create a inquirer callback
 *
 * @param {string} title
 * @param {object} prompt
 * @param {function} callback
 * @returns {object}
 *
 * @example
 * const answers = async () => await makeInquirer({
 *    title: "Adding licenses...",
 *    prompt: pubGenPrompt("addLicense"),
 * });
 */
export async function makeInquirer({
    title,
    prompt,
    callback,
    publication,
    type = "info",
}) {
    let packageJSON = {};
    try {
        packageJSON = packageJson(publication);
    } catch (error) {
        packageJSON = pkg;
    }
    const console = new Console(packageJSON);
    console.log(title, { type });
    if (typeof callback === "undefined") {
        callback = async (answers) => {
            return cleanFalsy(answers);
        };
    }
    return await inquirer.prompt(prompt).then(callback);
}

/**
 * Creates the README.md file for the publication
 *
 * @param {string} name | name of the publication
 * @param {Object} answers | package.json of the publication
 * @param {string} answers.description | description of the publication
 *
 * @returns | README.md as string (needs to be saved)
 */
export function createREADME(name, answers) {
    return toMD([
        `# \`${publicationPkgName(name)}\`\n`,
        `> Publicação gerada com [pub-gen](https://elucidario.art/pub-gen)`,
        answers.description,
    ]);
}

/**
 * Creates the .gitignore file for the publication
 *
 * @returns | .gitignore as string (needs to be saved)
 */
export function createGitignore(files = []) {
    return [
        "node_modules",
        "credentials.json",
        "token.json",
        "dist",
        "~$*.*",
        ...files,
    ].join("\n");
}

/**
 * Creates the pub-gen.json file for the publication
 * use for current version schema
 *
 * @see package.json->config.schema-version for current version
 * @see packages/pub-gen/static/pub-gen/schemas/pub-gen-schema.json for schema
 *
 * @param {string} name | name of the publication
 * @param {Object} answers | answers from the prompt
 *
 * @returns | pub-gen.json as object (needs to be stringified to be saved)
 */
export function createPubGenJson(name, answers) {
    const { keywords, ...rest } = answers;
    const version = pkg.config["schema-version"];
    return cleanFalsy({
        $schema:
            "node_modules/@elucidario/pkg-pub-gen/static/pub-gen/schemas/pub-gen-schema.json",
        version,
        profile: "data-package",
        id: `https://elucidario.art/publicacoes/${name.trim()}`,
        ...rest,
        keywords:
            typeof keywords !== "undefined"
                ? keywords.split(/[;,]/).map((k) => k.trim())
                : undefined,
        created: new Date().toISOString(),
    });
}

/**
 * Creates the package.json file for the publication
 * @param {string} name | name of the publication
 * @returns | package.json
 */
export function createPackageJson(name) {
    const pubName = kebabCase(name.trim());
    return {
        name: publicationPkgName(name),
        version: "0.1.0",
        main: "README.md",
        private: true,
        scripts: {
            "add-author": `pub-gen add-author -p ${pubName}`,
            authenticate: `pub-gen authenticate -p ${pubName}`,
            build: `pnpm clean && pub-gen build -p ${pubName}`,
            clean: "rm -rf dist/*",
            convert: `pub-gen convert -p ${pubName}`,
            "ref-add": `pub-gen reference add -p ${pubName}`,
            "ref-search": `pub-gen reference search -p ${pubName}`,
            "version-up": `pub-gen version -p ${pubName}`,
        },
        devDependencies: {
            "@elucidario/pkg-pub-gen": "workspace:^",
        },
    };
}

/**
 * Parse node value
 *
 * @param {Object} node
 * @returns {Object}
 */
export function parseNodeValue(value) {
    const [action, optionsString] = value
        .replace("{{", "")
        .replace("}}", "")
        .split(":");

    const [filePath, ...fileOptions] = optionsString.split(";");

    const options = {};
    for (let option of fileOptions) {
        const [key, value] = option.split("=");
        options[key] = value;
    }

    if (action === "count") {
        return {
            action,
            optionsString,
            type: filePath,
            options,
        };
    }

    return {
        action,
        optionsString,
        filePath,
        fileOptions: options,
    };
}

/**
 *  Check if node value is a pub-gen node value ({{...}})
 * @param {string} value | node value
 * @returns | true if node value is a pub-gen node value ({{...}}), false otherwise
 */
export function isPubGenNodeValue(value) {
    return value.startsWith("{{") && value.endsWith("}}");
}

/**
 * Replace Regex Handlebars
 *
 * @param {string} text
 * @param {Object} options
 * @returns {string}
 */
export function replaceRegexHandlebars(text, options) {
    if (!text) return null;
    const regex = /{{(.*?)}}/g;
    const replacedText = text.replace(regex, (match, group) => {
        const replacement = options[group.trim()] || "";
        return replacement;
    });

    return replacedText;
}

/**
 *
 * @param {string} md | Markdown string
 * @param {object} options | options for parser
 * @param {boolean} options.reduce | reduce mdast tree
 * @returns
 */
export function mdToMdast(md, options = undefined) {
    if (typeof options === "undefined" || !options.reduce)
        return fromMarkdown(md);

    return fromMarkdown(md).children.reduce((acc, node) => {
        switch (node.type) {
            case "root":
            case "paragraph":
                acc = node.children;
                break;
        }
        return acc;
    }, []);
}

/**
 * Flatten style structure files
 * @param {object} data
 * @param {array} exclude
 * @param {array} paths
 * @returns {array}
 */
export function flattenStyleStructureFiles(data, exclude = [], paths = []) {
    for (const [key, value] of Object.entries(data)) {
        if (typeof value === "string" && value.endsWith(".md")) {
            paths.push(value);
        } else if (Array.isArray(value)) {
            for (const [name, include] of value) {
                if (
                    include &&
                    !exclude.includes(name) &&
                    !exclude.includes(key)
                ) {
                    paths.push(`${key}/${name}.md`);
                }
            }
        } else if (typeof value === "object") {
            flattenStyleStructureFiles(value, exclude, paths); // Recursively traverse
        }
    }
    return paths;
}

/**
 * Loop through style structure
 * @param {object} structure
 * @param {function} callback
 * @param {string} prefix
 * @returns {array}
 */
export async function loopStylesStructure(structure, callback, prefix = "") {
    for (const [key, value, title] of Object.entries(structure)) {
        if (typeof value === "string") {
            await callback({
                key: prefix ? prefix : key,
                value,
                title,
            });
        } else if (Array.isArray(value)) {
            for (const [name, required, title] of value) {
                await callback({
                    key: prefix ? prefix : key,
                    value: `${prefix ? `${prefix}/` : ""}${key}/${name}`,
                    required,
                    title,
                });
            }
        } else if (typeof value === "object") {
            await loopStylesStructure(value, callback, key); // Recursively traverse
        }
    }
}

/**
 * Process files
 * @param {object} obj
 * @param {function} callback
 * @returns {object}
 */
export function processFiles(obj, callback) {
    if (Array.isArray(obj)) {
        return obj.map((item) => processFiles(item, callback));
    } else if (typeof obj === "object") {
        if (
            obj.hasOwnProperty("file") &&
            obj.hasOwnProperty("path") &&
            obj.hasOwnProperty("processed")
        ) {
            return callback(obj);
        } else {
            return Object.fromEntries(
                Object.entries(obj).map(([key, value]) => {
                    return [key, processFiles(value, callback)];
                }),
            );
        }
    } else {
        return obj;
    }
}
