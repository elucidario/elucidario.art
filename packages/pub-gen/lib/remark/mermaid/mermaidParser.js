import path from "path";
import * as unist from "@elucidario/pkg-unist";
import { parseNodeValue, isPubGenNodeValue, mdToMdast } from "../../utils";
import { getPaths, readFile } from "@elucidario/pkg-paths";
import parser from "parser-front-matter";
import { visit } from "unist-util-visit";
import { Console } from "@elucidario/pkg-console";

const mermaidParser = (treeOptions) => {
    const { publication, lang, assets, version, pkg, ...extra } = treeOptions;

    // default docxProcessor to pandoc while we don't implement remark-docx.
    let docxProcessor = "pandoc";
    if (typeof extra.docxProcessor !== "undefined")
        docxProcessor = extra.docxProcessor;

    const console = new Console(pkg);

    return (tree, file) => {
        const parsed = [];

        visit(tree, "text", async (node, index, parent) => {
            if (!isPubGenNodeValue(node.value)) return;

            const { action, filePath, fileOptions } = parseNodeValue(
                node.value,
            );

            if (action !== "mermaid") return;

            const mermaidData = readFile({
                filePath: path.resolve(
                    getPaths().publications,
                    publication,
                    "content",
                    lang,
                    filePath,
                ),
                ext: "md",
            }).content;

            let { data, content } = parser.parseSync(mermaidData);

            let {
                filename,
                source,
                title,
                width,
                theme,
                background,
                format,
                type,
            } = data;

            const mermaidOptions = {
                filename,
                loc: JSON.stringify(
                    path.resolve(
                        getPaths().publications,
                        publication,
                        "files",
                        "generated",
                    ),
                ),
                width,
                theme,
                background,
                format,
            };

            let pandocOptions = "{.mermaid ";
            pandocOptions += Object.keys(mermaidOptions)
                .map((key) =>
                    typeof mermaidOptions[key] !== "undefined"
                        ? `${key}=${mermaidOptions[key]}`
                        : false,
                )
                .filter(Boolean)
                .join(" ");
            pandocOptions += "}";

            content = content
                .replace(/```mermaid\r\n/, "")
                .replace(/```\r\n/, "")
                .trim();

            const codeAst = unist.code(
                content,
                docxProcessor === "pandoc" ? pandocOptions : "mermaid",
            );

            node.type = "code";
            node.value = codeAst.value;
            node.lang = codeAst.lang;

            parsed.push({
                node,
                parent,
                index,
                title,
                source,
            });
        });

        parsed.forEach((el) => {
            // find index of el.parent in tree
            const parentIndex = tree.children.findIndex(
                (child) =>
                    child.position.start.line === el.parent.position.start.line,
            );

            // insert title node before el.node in tree.
            tree.children.splice(
                parentIndex,
                0,
                unist.paragraph(mdToMdast(el.title, { reduce: true })),
            );

            // insert source node after el.node in tree.
            tree.children.splice(
                parentIndex + 2,
                0,
                unist.paragraph(mdToMdast(el.source, { reduce: true })),
            );
        });
    };
};

export default mermaidParser;
