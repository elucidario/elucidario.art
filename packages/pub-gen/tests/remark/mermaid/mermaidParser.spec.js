import path from "path";

import mermaidParser from "../../../lib/remark/mermaid/mermaidParser";
import { readFile, getPaths } from "@elucidario/pkg-paths";

import testProcessor from "../../mocks/remarkProcessor";
import { packageJson } from "../../../lib/utils";

describe("mermaidParser", () => {
    const mermaidContent = readFile(
        path.resolve(
            getPaths().publications,
            "publicacao-teste",
            "content",
            "pt-br",
            "internal",
            "body",
            "hello-world.md",
        ),
    ).value;

    const pkg = packageJson("publicacao-teste");

    it("should return mermaidParser node", async () => {
        const mermaidParserNode = await testProcessor(mermaidContent, {
            plugin: mermaidParser,
            pluginOptions: {
                publication: "publicacao-teste",
                lang: "pt-BR",
                pkg,
            },
            stringifyOptions: {},
            publication: "publicacao-teste",
            lang: "pt-BR",
        });
        expect(mermaidParserNode).toMatchSnapshot();
    });
});
