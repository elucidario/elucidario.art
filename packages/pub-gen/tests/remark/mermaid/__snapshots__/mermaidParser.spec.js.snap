// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mermaidParser should return mermaidParser node 1`] = `
VFile {
  "cwd": "C:\\Users\\55119\\Elucidário.art\\elucidario\\packages\\pub-gen",
  "data": {},
  "history": [],
  "messages": [],
  "value": "# Hello World

## Second level heading

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vitae diam eget risus var \\[@linked-art2021.1].

{{count:figures;legend=\\*\\*Mermaid\\*\\*}}

\`\`\`{.mermaid&#x20;filename=mermaid&#x20;loc="C:\\\\\\Users\\\\\\55119\\\\\\Elucidário.art\\\\\\elucidario\\\\\\publications\\\\\\publicacao-teste\\\\\\files\\\\\\generated"&#x20;width=720&#x20;background=transparent}
erDiagram
        OBJETO ||--|{ CLASSIFICACAO : ""
\`\`\`

**Fonte:** Elaborado pelo autor.

### Third level heading

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vitae diam eget risus var. \\[@linked-art2021.1, p. 2]

#### Fourth level heading

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vitae diam eget risus var. \\[@linked-art2021.2]

##### Fifth level heading

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vitae diam eget risus var. \\[@linked-art2021.2; @linked-art2021.1]

###### Sixth level heading

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed vitae diam eget risus var.

{{count:figures;legend=\\*\\*Lorem ipsum\\*\\*}}

![Lorem ipsum](/../../files/static/lorem-ipsum.jpeg)
",
}
`;
